src/inbox/rx.rs

```diff
@@ -0,0 +1,303 @@
+        }
+
+        None
+    }
+}
+
+impl<A, Rc: RxRefCounter> Drop for ReceiveFuture<A, Rc> {
+    fn drop(&mut self) {
+        if let ReceiveFutureInner::Waiting { waiting, rx } =
+            mem::replace(&mut self.0, ReceiveFutureInner::Complete)
+        {
+            let mut inner = rx.inner.chan.lock().unwrap();
+
+            // This receive future was woken with a message - send in back into the channel.
+            // Ordering is compromised somewhat when concurrent receives are involved, and the cap
+            // may overflow (probably not enough to cause backpressure issues), but this is better
+            // than dropping a message.
+            if let Some(WakeReason::StolenMessage(msg)) = waiting.lock().cancel() {
+                match inner.try_fulfill_receiver(WakeReason::StolenMessage(msg)) {
+                    Ok(()) => (),
+                    Err(WakeReason::StolenMessage(msg)) => {
+                        if msg.priority == Priority::default() {
+                            // Preserve ordering as much as possible by pushing to the front
+                            inner.ordered_queue.push_front(msg.val)
+                        } else {
+                            inner.priority_queue.push(msg);
+                        }
+                    }
+                    Err(_) => unreachable!("Got wrong wake reason back from try_fulfill_receiver"),
+                }
+            }
+        }
+    }
+}
+
+impl<A, Rc: RxRefCounter> FusedFuture for ReceiveFuture<A, Rc> {
+    fn is_terminated(&self) -> bool {
+        matches!(self.0, ReceiveFutureInner::Complete)
+    }
+}
+
+pub struct WaitingReceiver<A> {
+    waker: Option<Waker>,
+    wake_reason: Option<WakeReason<A>>,
+}
+
+impl<A> Default for WaitingReceiver<A> {
+    fn default() -> Self {
+        WaitingReceiver {
+            waker: None,
+            wake_reason: None,
+        }
+    }
+}
+
+impl<A> WaitingReceiver<A> {
+    pub(super) fn fulfill(&mut self, reason: WakeReason<A>) -> Result<(), WakeReason<A>> {
+        match self.wake_reason {
+            // Receive was interrupted, so this cannot be fulfilled
+            Some(WakeReason::Cancelled) => return Err(reason),
+            Some(_) => unreachable!("Waiting receiver was fulfilled but not popped!"),
+            None => self.wake_reason = Some(reason),
+        }
+
+        if let Some(waker) = self.waker.take() {
+            waker.wake();
+        }
+
+        Ok(())
+    }
+
+    /// Signify that this waiting receiver was cancelled through [`ReceiveFuture::cancel`]
+    fn cancel(&mut self) -> Option<WakeReason<A>> {
+        mem::replace(&mut self.wake_reason, Some(WakeReason::Cancelled))
+    }
+}
+
+pub(crate) trait RxRefCounter: Unpin {
+    fn increment<A>(&self, inner: &Chan<A>) -> Self;
+    #[must_use = "If decrement returns false, the address must be disconnected"]
+    fn decrement<A>(&self, inner: &Chan<A>) -> bool;
+}
+
+pub(crate) struct RxStrong(pub(crate) ());
```

Comment:
The inner is to prevent it from being constructed by users for semver purposes. I think it needs to be pub(crate) for it to be constructed where it needs to be, but it might be able to be reduced to pub(in (some module)) or similar

Reply:
Right, because it is re-exported!

I think I got confused because it is declared as `pub(crate)` and didn't expect it to be public. Perhaps make it `pub struct RxStrong` to make it more obvious that this is a public type?

Regarding the field, I would suggest to make it completely private and perhaps a `pub(crate)` constructor?

Reply:
I think RxStrong itself is not re-exported, actually. It might be worth keeping the inner field around for encapsulation purposes - plus, it's a little harder to accidentally construct a new one by `RxStrong(())` than by `RxStrong`, I think.

Reply:
I've pushed https://github.com/Restioson/xtra/pull/85/commits/de52e4773abbd2c8df0803c42eeedabb10101bdf which tries to accomplish this